### ISSUE TITLE
[Data] read_text fails to read files from S3 when setting ignore_missing_paths to True

### ISSUE THREAD

**Issue by nate-bush** [ray-project/ray#41377]
### What happened + What you expected to happen

`read_text` fails to read files from S3 when setting `ignore_missing_paths` to True. I get the following error:

```SHELL
---------------------------------------------------------------------------
RayTaskError(ValueError)                  Traceback (most recent call last)
Cell In[69], line 8
----> 8 dataset = ray.data.read_text(
      9     paths=s3_files,
     10     ignore_missing_paths=True,
     11 )

File ~/anaconda3/lib/python3.8/site-packages/ray/data/read_api.py:1302, in read_text(paths, encoding, drop_empty_lines, filesystem, parallelism, ray_remote_args, arrow_open_stream_args, meta_provider, partition_filter, partitioning, ignore_missing_paths)
   1300 if meta_provider is None:
   1301     meta_provider = get_generic_metadata_provider(TextDatasource._FILE_EXTENSION)
-> 1302 return read_datasource(
   1303     TextDatasource(),
   1304     parallelism=parallelism,
   1305     paths=paths,
   1306     filesystem=filesystem,
   1307     ray_remote_args=ray_remote_args,
   1308     open_stream_args=arrow_open_stream_args,
   1309     meta_provider=meta_provider,
   1310     partition_filter=partition_filter,
   1311     partitioning=partitioning,
   1312     drop_empty_lines=drop_empty_lines,
   1313     encoding=encoding,
   1314     ignore_missing_paths=ignore_missing_paths,
   1315 )

File ~/anaconda3/lib/python3.8/site-packages/ray/_private/auto_init_hook.py:24, in wrap_auto_init.<locals>.auto_init_wrapper(*args, **kwargs)
     21 @wraps(fn)
     22 def auto_init_wrapper(*args, **kwargs):
     23     auto_init_ray()
---> 24     return fn(*args, **kwargs)

File ~/anaconda3/lib/python3.8/site-packages/ray/data/read_api.py:371, in read_datasource(datasource, parallelism, ray_remote_args, **read_args)
    363     scheduling_strategy = NodeAffinitySchedulingStrategy(
    364         ray.get_runtime_context().get_node_id(),
    365         soft=False,
    366     )
    367     get_reader = cached_remote_fn(
    368         _get_reader, retry_exceptions=False, num_cpus=0
    369     ).options(scheduling_strategy=scheduling_strategy)
--> 371     (requested_parallelism, min_safe_parallelism, inmemory_size, reader,) = ray.get(
    372         get_reader.remote(
    373             datasource,
    374             ctx,
    375             cur_pg,
    376             parallelism,
    377             local_uri,
    378             _wrap_arrow_serialization_workaround(read_args),
    379         )
    380     )
    382 # TODO(hchen/chengsu): Remove the duplicated get_read_tasks call here after
    383 # removing LazyBlockList code path.
    384 read_tasks = reader.get_read_tasks(requested_parallelism)

File ~/anaconda3/lib/python3.8/site-packages/ray/_private/auto_init_hook.py:24, in wrap_auto_init.<locals>.auto_init_wrapper(*args, **kwargs)
     21 @wraps(fn)
     22 def auto_init_wrapper(*args, **kwargs):
     23     auto_init_ray()
---> 24     return fn(*args, **kwargs)

File ~/anaconda3/lib/python3.8/site-packages/ray/_private/client_mode_hook.py:103, in client_mode_hook.<locals>.wrapper(*args, **kwargs)
    101     if func.__name__ != "init" or is_client_mode_enabled_by_default:
    102         return getattr(ray, func.__name__)(*args, **kwargs)
--> 103 return func(*args, **kwargs)

File ~/anaconda3/lib/python3.8/site-packages/ray/_private/worker.py:2547, in get(object_refs, timeout)
   2545     worker.core_worker.dump_object_store_memory_usage()
   2546 if isinstance(value, RayTaskError):
-> 2547     raise value.as_instanceof_cause()
   2548 else:
   2549     raise value

RayTaskError(ValueError): ray::_get_reader() (pid=1248, ip=10.0.5.211)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/data/read_api.py", line 2348, in _get_reader
    reader = ds.create_reader(**kwargs)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/data/datasource/file_based_datasource.py", line 256, in create_reader
    return _FileBasedDatasourceReader(self, **kwargs)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/data/datasource/file_based_datasource.py", line 486, in __init__
    self._paths, self._file_sizes = map(
ValueError: not enough values to unpack (expected 2, got 0)
```

Note that this also applied to `read_json`

### Versions / Dependencies

ray @ file:///home/ray/ray-2.7.1-cp38-cp38-manylinux2014_x86_64.whl#sha256=afbb656eac81bf9a3b22ab4fd0ef2605c6a92768ed86333322d8c6003effe3a9

### Reproduction script

```PYTHON
s3_files = [...list of s3 URLs where a file is missing...]
dataset = ray.data.read_text(
    paths=s3_files,
    ignore_missing_paths=True,
)
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by wingkitlee0** [comment]
FYI, it runs fine on local files
```
Python 3.8.18 (default, Nov 14 2023, 22:56:19)
Type 'copyright', 'credits' or 'license' for more information
IPython 8.12.3 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import ray

In [2]: ray.__version__
Out[2]: '2.7.1'

In [3]: ray.data.read_text(["a", "b", "c"], ignore_missing_paths=True).take_all()
...
Out[3]: [{'text': 'this is a cat.'}, {'text': 'this is not a cat.'}]
```

**Comment by nate-bush** [comment]
@wingkitlee0 Good call out. I reproduced on my side that this bug doesn't apply to local files. I'll update the ticket to reflect that fact.

**Comment by scottjlee** [comment]
@nate-bush do you have a complete reproducible example, including s3 paths which we can use? I tried a similar approach as the one you described above, but cannot reproduce the error on ray 2.7.1 or ray master:

```
>>> import ray
>>> paths = [
...     "s3://air-example-data/sql/train.jsonl",
...     "s3://air-example-data/sql/missing.jsonl",
...     "s3://air-example-data/missing-directory"
... ]

# Also tried this to artificially increase the number of files to read, no issues.
>>> # paths = paths * 50

>>> ds = ray.data.read_text(paths, ignore_missing_paths=True)
Dataset(num_blocks=20, num_rows=30000, schema={text: string})
```

**Comment by nate-bush** [comment]
Thanks @scottjlee. We had a couple moving parts so it's hard for me to replicate my issue. 

1. We recently reconfigured out Anyscale platform and after AWS IAM changes.
2. We completely changed our S3 permissions.
3. The data I'm querying is different. 

I think the problem was either (1) a permission issue that was resolved or (2) none of the provided S3 paths existed - I think if every path is missing then it results in an error.
